src/fuzz_introspector/frontends/src/frontend_go.py

```diff
@@ -0,0 +1,610 @@
+
+
+class SourceCodeFile():
+    """Class for holding file-specific information."""
+
+    def __init__(self,
+                 source_file: str,
+                 source_content: Optional[bytes] = None):
+        logger.info('Processing %s' % source_file)
+
+        self.root = None
+        self.imports = []
+        self.source_file = source_file
+        self.tree_sitter_lang = Language(tree_sitter_go.language())
+        self.parser = Parser(self.tree_sitter_lang)
+
+        if source_content:
+            self.source_content = source_content
+        else:
+            with open(self.source_file, 'rb') as f:
+                self.source_content = f.read()
+
+        # List of function definitions in the source file.
+        self.functions = []
+        self.methods = []
+
+        # Initialization ruotines
+        self.load_tree()
+
+        # Load function/method declaration
+        self._set_function_declaration()
+        self._set_method_declaration()
+
+        # Parse import package
+        self._set_imports()
+
+    def load_tree(self):
+        """Load the the source code into a treesitter tree, and set
+        the root node."""
+        self.root = self.parser.parse(self.source_content).root_node
+
+    def _set_function_declaration(self):
+        """Internal helper for retrieving all functions."""
+        func_query_str = '( function_declaration ) @fd '
+        func_query = self.tree_sitter_lang.query(func_query_str)
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.functions.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, True))
+
+    def _set_method_declaration(self):
+        """Internal helper for retrieving all methods."""
+        func_query_str = '( method_declaration ) @fd '
+        func_query = self.tree_sitter_lang.query(func_query_str)
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.methods.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, False))
+
+    def _set_imports(self):
+        """Internal helper for retrieving all imports."""
+        import_pattern = r'^\s*import\s+(?:(?P<single>"[^"]+")|\((?P<multi>[^)]+)\))'
+
+        imports = []
+        for match in re.finditer(import_pattern,
+                                 self.source_content.decode('utf8'),
+                                 re.MULTILINE):
+            if match.group('single'):
+                imports.append(match.group('single'))
+            elif match.group('multi'):
+                multi = [
+                    line.strip().strip('"')
+                    for line in match.group('multi').splitlines()
+                ]
+                imports.extend(multi)
+
+        self.imports = [
+            s.rsplit('/', 1)[-1] if '/' in s else s for s in imports
```

Comment:
could you leave a comment on this? What does it do

Reply:
Done so.